plugins/modules/keycloak_user_federation.py

```diff
@@ -347,6 +347,7 @@
                       to the first part of his Kerberos principal. For instance, for principal C(john@KEYCLOAK.ORG),
                       it will assume that LDAP username is V(john).
                 type: str
+                default: 'userPrincipalName'
```

Comment:
Changing default values are breaking changes. Also generally default values should be avoided, the behavior of the module should be "if no value is provided, do not change it". Having a default makes that behavior impossible, and instead forces "if no value is provided, change the configuration to the default set here", which is often not what users want.

Reply:
OKay, makes sense. Changing the default is probably not the best way.

My problem is, that the module always detects a change, because the API sets the value ``userPrincipalName`` when the parameter is unset in the desired config when creating a user federation (when updating with the unset parameter the value remains unchanged). So we have always ``krbPrincipalAttribute = userPrincipalName`` vs unset. 
The other problem with the parameter is, that the API completely removes the parameter if it is set to an emtpy string. SO again the module always detects a change (empty string vs unset).

I will test and think about it some more and move the other changes to separate PRs.

Reply:
I guess then the module shouldn't try to change something that the user didn't provide?

Reply:
No, i guess not. 

Most of the other changes are already included in other PRs, so i will remove this one.